staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/cel/validation_test.go

```diff
@@ -0,0 +1,1227 @@
+					},
+					"spec": map[string]interface{}{
+						"field1": "a",
+					},
+				},
+			},
+			schema: objectTypePtr(map[string]schema.Structural{
+				"embedded": {
+					Generic: schema.Generic{Type: "object"},
+					Extensions: schema.Extensions{
+						XEmbeddedResource: true,
+					},
+					Properties: map[string]schema.Structural{
+						"kind":       stringType,
+						"apiVersion": stringType,
+						"metadata": objectType(map[string]schema.Structural{
+							"name":         stringType,
+							"generateName": stringType,
+						}),
+						"spec": objectType(map[string]schema.Structural{
+							"field1": stringType,
+						}),
+					},
+				},
+			}),
+			valid: []string{
+				"self.embedded.kind == 'Pod'",
+				"self.embedded.apiVersion == 'v1'",
+				"self.embedded.metadata.name == 'foo'",
+				"self.embedded.metadata.generateName == 'pickItForMe'",
+				"self.embedded.spec.field1 == 'a'",
+			},
+			errors: map[string]string{
+				// only name and generateName are accessible on metadata
+				"has(self.embedded.metadata.namespace)": "undefined field 'namespace'",
+			},
+		},
+		{name: "embedded object with prune unknown",
+			obj: map[string]interface{}{
+				"embedded": map[string]interface{}{
+					"apiVersion": "v1",
+					"kind":       "Pod",
+					"metadata": map[string]interface{}{
+						"name":         "foo",
+						"generateName": "pickItForMe",
+						"namespace":    "xyz",
+					},
+					"spec": map[string]interface{}{
+						"field1": "a",
+					},
+				},
+			},
+			schema: objectTypePtr(map[string]schema.Structural{
+				"embedded": {
+					Generic: schema.Generic{Type: "object"},
+					Extensions: schema.Extensions{
+						XPreserveUnknownFields: true,
+						XEmbeddedResource:      true,
+					},
+				},
+			}),
+			valid: []string{
+				// access to metadata allowed even if not specified in schema
+				// because the accessible fields (name and generateName) are fixed
+				"self.embedded.kind == 'Pod'",
+				"self.embedded.apiVersion == 'v1'",
+				"self.embedded.metadata.name == 'foo'",
+				"self.embedded.metadata.generateName == 'pickItForMe'",
+				"has(self.embedded)",
+			},
+			// only field declared in the schema can be field selected in CEL expressions
+			errors: map[string]string{
+				"has(self.embedded.spec)": "undefined field 'spec'",
+			},
+		},
+		{name: "string in intOrString",
+			obj: map[string]interface{}{
+				"something": map[string]interface{}{
+					"strVal": "25%",
+				},
+			},
+			schema: objectTypePtr(map[string]schema.Structural{
+				"something": intOrStringType(),
+			}),
+			valid: []string{"type(self.something) == string && self.something == '25%'"},
+		},
+		{name: "int in intOrString",
+			obj: map[string]interface{}{
+				"something": map[string]interface{}{
+					"intVal": 1,
+				},
+			},
+			schema: objectTypePtr(map[string]schema.Structural{
+				"something": intOrStringType(),
+			}),
+			valid: []string{"type(self.something) == int && self.something == 1"},
```

Comment:
and this works now because the type is dynamic? I.e. `self.something == ["foo"|` would type fine too?

Reply:
Since it's dynamic, `something` is treated as the union of all types at compile time. But at runtime type checking happens according to the usual rules. So `self.something == ["foo"]` would compile but fail with a `no such overload for _==_ (string, list)` or `no such overload for _==_ (int, list)` depending on what type the value takes at runtime.

CEL go support for unions is slated for this year though. I've discussed it with @TristonianJones and it will make this more strict (`self.something == ["foo"]` would fail at compile time), I've listed that in the graduation criteria for beta.

Reply:
I've added some error test cases to describe this behavior in more detail.

Reply:
Oh, actually, it all should be in the schemas code (since there is recursion that happens there). Fixed.